crates/src/rating/elo.cairo

```diff
@@ -0,0 +1,174 @@
+//! Elo struct and rating methods.
+//! Source: https://github.com/saucepoint/elo-lib/blob/main/src/Elo.sol
+
+// Core imports
+
+use core::integer::{u32_sqrt, u64_sqrt, u128_sqrt, u256_sqrt};
+
+// Constants
+
+const MULTIPLIER: u32 = 10_000;
+const SCALER: u256 = 800;
+
+// Errors
+
+mod errors {
+    const ELO_DIFFERENCE_TOO_LARGE: felt252 = 'Elo: difference too large';
+}
+
+/// Elo implementation..
+#[generate_trait]
+impl EloImpl of EloTrait {
+    /// Calculates the change in ELO rating, after a given outcome.
+    /// # Arguments
+    /// * `rating_a` - The ELO rating of the player A.
+    /// * `rating_b` - The ELO rating of the player B.
+    /// * `score` - The score of the player A, scaled by 100. 100 = win, 50 = draw, 0 = loss.
+    /// * `k` - The k-factor or development multiplier used to calculate the change in ELO rating. 20 is the typical value.
+    /// # Returns
+    /// * `change` - The change in ELO rating of player A.
+    /// * `negative` - The directional change of player A's ELO. Opposite sign for player B.
+    fn rating_change<
+        T,
+        +Sub<T>,
+        +PartialOrd<T>,
+        +Into<T, u256>,
+        +Drop<T>,
+        +Copy<T>,
+        S,
+        +Into<S, u32>,
+        +Drop<S>,
+        +Copy<S>,
+        K,
+        +Into<K, u32>,
+        +Drop<K>,
+        +Copy<K>,
+        C,
+        +Into<u32, C>,
+    >(
+        rating_a: T, rating_b: T, score: S, k: K
+    ) -> (C, bool) {
+        let negative = rating_a > rating_b;
+        let rating_diff: u256 = if negative {
```

Comment:
If we can use signed ints here, this will be a lot simpler

Reply:
Done, not that it seems to be unsafe to do:

```cairo
let order_felt: felt252 = (SCALER + rating_diff).into();
let order: u256 = order_felt.into() / 25;
```

But actually rating_diff is verified to be `> -800` earlier so the result cannot be negative